File: packages/typescript/src/features/advanced-capabilities/clipboard/actions/write.ts
import {
	type Action,
	type HandlerCallback,
	type HandlerOptions,
	type IAgentRuntime,
	logger,
	type Memory,
	ModelType,
	parseKeyValueXml,
	type State,
} from "../../../../types/index.ts";
import { createClipboardService } from "../services/clipboardService.ts";
import { requireActionSpec } from "../specs.ts";

interface WriteInput {
	title: string;
	content: string;
	tags?: string[];
}

function isValidWriteInput(obj: Record<string, unknown>): boolean {
	return (
		typeof obj.title === "string" &&
		obj.title.length > 0 &&
		typeof obj.content === "string" &&
		obj.content.length > 0
	);
}

const EXTRACT_TEMPLATE = `Extract the following information from the user's message to save to the clipboard:

User message: {{text}}

Recent conversation:
{{messageHistory}}

Respond with XML containing:
- title: A short, descriptive title for the note (required)
- content: The main content to save (required)
- tags: Comma-separated tags for categorization (optional)

<response>
<title>The note title</title>
<content>The content to save</content>
<tags>tag1, tag2</tags>
</response>`;

async function extractWriteInfo(
	runtime: IAgentRuntime,
	message: Memory,
	_state: State,
): Promise<WriteInput | null> {
	const prompt = EXTRACT_TEMPLATE.replace(
		"{{text}}",
		message.content.text ?? "",
	).replace("{{messageHistory}}", "");

	const result = await runtime.useModel(ModelType.TEXT_SMALL, {
		prompt,
		stopSequences: [],
	});

	logger.debug("[ClipboardWrite] Extract result:", result);

	const parsed = parseKeyValueXml(String(result)) as Record<
		string,
		unknown
	> | null;

	if (!parsed || !isValidWriteInput(parsed)) {
		logger.error("[ClipboardWrite] Failed to extract valid write info");
		return null;
	}

	const tags = parsed.tags
		? String(parsed.tags)
				.split(",")
				.map((t: string) => t.trim())
				.filter(Boolean)
		: undefined;

	return {
		title: String(parsed.title),
		content: String(parsed.content),
		tags,
	};
}

const spec = requireActionSpec("CLIPBOARD_WRITE");

export const clipboardWriteAction: Action = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		options?: HandlerOptions,
	): Promise<boolean> => {
		const __avTextRaw =
			typeof message?.content?.text === "string" ? message.content.text : "";
		const __avText = __avTextRaw.toLowerCase();
		const __avKeywords = ["clipboard", "write"];
		const __avKeywordOk =
			__avKeywords.length > 0 &&
			__avKeywords.some((kw) => kw.length > 0 && __avText.includes(kw));
		const __avRegex = /\b(?:clipboard|write)\b/i;
		const __avRegexOk = __avRegex.test(__avText);
		const __avSource = String(message?.content?.source ?? "");
		const __avExpectedSource = "";
		const __avSourceOk = __avExpectedSource
			? __avSource === __avExpectedSource
			: Boolean(__avSource || state || runtime?.agentId || runtime?.getService);
		const __avOptions = options && typeof options === "object" ? options : {};
		const __avInputOk =
			__avText.trim().length > 0 ||
			Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
			Boolean(message?.content && typeof message.content === "object");

		if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
			return false;
		}

		const __avLegacyValidate = async (
			_runtime: IAgentRuntime,
			message: Memory,
		): Promise<boolean> => {
			// Check for clipboard-related intent in the message
			const text = (message.content?.text ?? "").toLowerCase();
			const hasSaveIntent =
				text.includes("save") ||
				text.includes("note") ||
				text.includes("remember") ||
				text.includes("write") ||
				text.includes("clipboard") ||
				text.includes("jot down") ||
				text.includes("store");

			return hasSaveIntent;
		};
		try {
			return Boolean(await __avLegacyValidate(runtime, message));
		} catch {
			return false;
		}
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		stateFromTrigger: State | undefined,
		_options: HandlerOptions | undefined,
		callback?: HandlerCallback,
		_responses?: Memory[],
	) => {
		const state = stateFromTrigger ?? (await runtime.composeState(message, []));
		const writeInfo = await extractWriteInfo(runtime, message, state);

		if (!writeInfo) {
			if (callback) {
				await callback({
					text: "I couldn't understand what you want me to save. Please provide a clear title and content for the note.",
					actions: ["CLIPBOARD_WRITE_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to extract write info" };
		}

		try {
			const service = createClipboardService(runtime);
			const entry = await service.write(writeInfo.title, writeInfo.content, {
				tags: writeInfo.tags,
			});

			const successMessage = `I've saved a note titled "${entry.title}" (ID: ${entry.id}).${
				entry.tags?.length ? ` Tags: ${entry.tags.join(", ")}` : ""
			} You can retrieve it later using the ID or by searching for it.`;

			if (callback) {
				await callback({
					text: successMessage,
					actions: ["CLIPBOARD_WRITE_SUCCESS"],
					source: message.content.source,
				});
			}

			return { success: true, text: successMessage, entryId: entry.id };
		} catch (error) {
			const errorMsg = error instanceof Error ? error.message : String(error);
			logger.error("[ClipboardWrite] Error:", errorMsg);
			if (callback) {
				await callback({
					text: `Failed to save the note: ${errorMsg}`,
					actions: ["CLIPBOARD_WRITE_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to write to clipboard" };
		}
	},

	examples: [],
};

export default clipboardWriteAction;

File: packages/python/elizaos/features/advanced_capabilities/clipboard/actions/write.py
"""Clipboard write action -- save content to the clipboard."""

from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Action, ActionResult, Content, ModelType
from elizaos.utils.xml import parse_key_value_xml

from ..services.clipboard_service import create_clipboard_service

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

EXTRACT_TEMPLATE = """Extract the following information from the user's message to save to the clipboard:

User message: {text}

Respond with XML containing:
- title: A short, descriptive title for the note (required)
- content: The main content to save (required)
- tags: Comma-separated tags for categorization (optional)

<response>
<title>The note title</title>
<content>The content to save</content>
<tags>tag1, tag2</tags>
</response>"""


async def _validate(
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | None = None,
) -> bool:
    text = (message.content.text or "").lower()
    save_keywords = ["save", "note", "remember", "write", "clipboard", "jot down", "store"]
    return any(kw in text for kw in save_keywords) and "clipboard" in text


async def _handler(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    callback: HandlerCallback | None = None,
    responses: list[Memory] | None = None,
) -> ActionResult:
    prompt = EXTRACT_TEMPLATE.format(text=message.content.text or "")
    result = await runtime.use_model(ModelType.TEXT_SMALL, prompt=prompt)
    parsed = parse_key_value_xml(str(result))

    if (
        not parsed
        or not isinstance(parsed.get("title"), str)
        or not isinstance(parsed.get("content"), str)
    ):
        if callback:
            await callback(
                Content(
                    text="I couldn't understand what you want me to save. Please provide a clear title and content.",
                    actions=["CLIPBOARD_WRITE_FAILED"],
                )
            )
        return ActionResult(text="Failed to extract write info", success=False)

    tags = None
    if parsed.get("tags"):
        tags = [t.strip() for t in str(parsed["tags"]).split(",") if t.strip()]

    try:
        service = create_clipboard_service(runtime)
        from ..types import ClipboardWriteOptions

        entry = await service.write(
            str(parsed["title"]),
            str(parsed["content"]),
            ClipboardWriteOptions(tags=tags),
        )
        msg = (
            f'I\'ve saved a note titled "{entry.title}" (ID: {entry.id}).'
            + (f" Tags: {', '.join(entry.tags)}" if entry.tags else "")
            + " You can retrieve it later using the ID or by searching for it."
        )
        if callback:
            await callback(Content(text=msg, actions=["CLIPBOARD_WRITE_SUCCESS"]))
        return ActionResult(text=msg, success=True, values={"entryId": entry.id})
    except Exception as e:
        if callback:
            await callback(
                Content(text=f"Failed to save the note: {e}", actions=["CLIPBOARD_WRITE_FAILED"])
            )
        return ActionResult(text="Failed to write to clipboard", success=False)


clipboard_write_action = Action(
    name="CLIPBOARD_WRITE",
    similes=["SAVE_NOTE", "WRITE_NOTE", "CLIPBOARD_SAVE"],
    description="Save content to the clipboard as a named note",
    validate=_validate,
    handler=_handler,
    examples=[],
)
